nbs/00-utils.ipynb

```diff
@@ -0,0 +1,188 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#default_exp utils"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Utility Functions\n",
+    "\n",
+    "> Utility functions to help with downstream tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#hide\n",
+    "from nbdev.showdoc import *\n",
+    "from self_supervised.byol import *\n",
+    "from self_supervised.simclr import *\n",
+    "from self_supervised.swav import *"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#export\n",
+    "from fastai.vision.all import *"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Loading Weights for Downstream Tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#export\n",
+    "def transfer_weights(learn:Learner, weights_path:Path, device:torch.device=None):\n",
+    "    \"Load and freeze pretrained weights inplace from `weights_path` using `device`\"\n",
+    "    if device is None: device = learn.dls.device\n",
+    "    new_state_dict = torch.load(weights_path, map_location=device)\n",
+    "    if 'model' in new_state_dict.keys(): new_state_dict = new_state_dict['model'] \n",
+    "    #allow for simply exporting the raw PyTorch model\n",
+    "    learn_state_dict = learn.model.state_dict()\n",
+    "    for name, param in learn_state_dict.items():\n",
+    "        name = 'encoder.'+name[2:]\n",
+    "        if name in new_state_dict:\n",
+    "            input_param = new_state_dict[name]\n",
+    "            if input_param.shape == param.shape:\n",
+    "                param.copy_(input_param)\n",
+    "            else:\n",
+    "                print('Shape mismatch at:', name, 'skipping')\n",
+    "        else: pass # these are weights that weren't in the original model, such as a new head\n",
+    "    learn.model.load_state_dict(learn_state_dict)\n",
+    "    learn.freeze()\n",
+    "    print(\"Weights successfully transferred!\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When training models with this library, the `state_dict` will change, so loading it back into `fastai` as an encoder won't be a perfect match. This helper function aims to make that simple. \n",
+    "\n",
+    "Example usage:\n",
+    "\n",
+    "First prepare the downstream-task dataset (`ImageWoof` is shown here):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "path = untar_data(URLs.IMAGEWOOF)\n",
```

Comment:
Can we modularize this as a function? You can use `get_dls()` from examples if you like.

Reply:
👍